src/imitation/rewards/reward_net.py

```diff
@@ -148,25 +165,38 @@ def __init__(
         observation_space: gym.Space,
         action_space: gym.Space,
         *,
-        scale: bool = False,
         discount_factor: float = 0.99,
+        **kwargs,
     ):
-        super().__init__(observation_space, action_space, scale=scale)
+        super().__init__(observation_space, action_space, **kwargs)
         self._discount_factor = discount_factor
 
         with tf.variable_scope("phi_network"):
             res = self.build_phi_network(self.obs_inp, self.next_obs_inp)
+            # end_potential is the potential when the episode terminates.
+            if discount_factor == 1.0:
+                # If undiscounted, terminal state must have potential 0.
+                end_potential = tf.constant(0.0)
+            else:
+                # Otherwise, it can be arbitrary, so make a trainable variable.
+                end_potential = tf.Variable(
+                    name="end_phi", shape=(), dtype=tf.float32, initial_value=0.0
+                )
```

Comment:
Why train an arbitrary potential for the absorbing state instead of just using zero?

Seems like if this value is positive, then it could encourage early episode termination in non-fixed-length environment.

Reply:
Normally potential shaping in the discounted case can represent adding a constant to the reward. If $\Phi(s) = k$ then each reward is increased by $(\gamma - 1)k$. If we fix potential shaping to be zero at the absorbing state it's removing the ability to add constant offsets to potential, therefore restricting the class of potential shaping we can represent.

It's plausible we don't need this degree of freedom and restricting it will make it easier to learn -- but it's just one extra parameter so I'd be surprised if it makes much difference.

In variable-horizon environments being able to learn to add a positive or negative constant to the reward seems useful. Sure, if it learns the wrong sign it'll hurt performance; but presumably it'd tend to learn the right sign? There's still the issue that episode termination leaks information so any results are dubious for evaluation, but I don't see a way to fix that.